src/Descriptor.cpp

```diff
@@ -151,3 +158,53 @@ void Descriptor::show_next_page(std::string_view input) noexcept {
         page_outbuf_.pop_front();
     }
 }
+
+bool Descriptor::try_start_snooping(Descriptor &other) {
+    // If already snooping, early out (to prevent us complaining it's a "snoop loop".
+    if (snooping_.count(&other) == 1)
+        return true;
+
+    struct SnoopLoopChecker {
+        std::unordered_set<Descriptor *> already_seen;
+        explicit SnoopLoopChecker(Descriptor *d) { already_seen.emplace(d); }
+        bool snoops(Descriptor *d) {
+            if (!already_seen.emplace(d).second)
+                return true; // if we didn't add a new entry to the set, we hit a duplicate!
+            return std::any_of(begin(d->snooping_), end(d->snooping_), [&](auto d) { return snoops(d); });
+        }
+    };
+    SnoopLoopChecker slc(this);
+    if (slc.snoops(&other))
+        return false;
+
+    other.snoop_by_.emplace(this);
+    snooping_.emplace(&other);
+    return true;
+}
+
+void Descriptor::stop_snooping(Descriptor &other) {
+    other.snoop_by_.erase(this);
+    snooping_.erase(&other);
+}
+
+void Descriptor::stop_snooping() {
+    while (!snooping_.empty())
+        stop_snooping(**snooping_.begin());
+}
+
+void Descriptor::close() noexcept {
+    while (!snoop_by_.empty()) {
+        auto &snooper = *snoop_by_.begin();
+        snooper->write("Your victim ({}) has left the game.\n\r"_format(character ? character->name : "unknown"));
+        snooper->stop_snooping(*this);
+    }
+    stop_snooping();
+}
+
```

Comment:
I think there may alternatives for 'note_input' that could be clearer, e.g. notify_snoopers? Your call.

Reply:
Yeah, I went back and forth here. `notify_snoopers()` is defininetly what it does _now_, but at some point in the future I can see that other stuff could be done here (logging maybe?). So I hedged my bets as if it were a general purpose "listener". For now I think I'll keep that; and then if we come across it again and it's still only doing snooping we can rename. thanks!